### PR TITLE
Rule edge cases

### PR DESCRIPTION
**Pull request by allthatilk**

### CODE REVIEW

**Comment by hayfield** [comment]
Examples of documentation style: http://www.sphinx-doc.org/en/stable/ext/example_google.html

File: iati/core/resources/test_data/202/valid_dateorder.xml [hunk 1]
```diff
@@ -4,5 +4,9 @@
   <planned-disbursement type="1">
    <period-start iso-date="2014-01-01" /><!-- Dates are in the correct order -->
    <period-end iso-date="2014-12-31" />
+   <period-start iso-date="2017-01-01" /><!-- Dates are in the correct order -->
+   <period-end iso-date="NOW" />
+   <period-start iso-date="2014-01-01+5:00" /><!-- Dates are in the correct order but have timezone values -->
+   <period-end iso-date="2014-12-31+5:00" />
```
Comment (hayfield) [inline iati/core/resources/test_data/202/valid_dateorder.xml]: Could also cover:

* Negative timezones
* Timezones that aren't on the hour
* `Z` rather than the `[\+-]\d?\d:\d\d` format

File: iati/core/resources/test_data/202/valid_unique.xml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0"?>
 
 <root_element>
-  <iati-identifier>This is a unique identifier</iati-identifier><!-- iati-identifier does not replicate -->
+  <iati-identifier>This is a unique identifier</iati-identifier><!-- elements do not replicate -->
```
Comment (hayfield) [inline iati/core/resources/test_data/202/valid_unique.xml]: The mix of IATI-specific and descriptive-for-the-test element names is somewhat confusing.

File: iati/core/resources/test_data/202/valid_unique.xml [hunk 3]
```diff
@@ -1,5 +1,6 @@
...
+  <other_unique_element></other_unique_element>
```
Comment (hayfield) [inline iati/core/resources/test_data/202/valid_unique.xml]: This is testing only elements - what about attributes?

(presume this is a wider thing - both elements and attributes should be tested both positively and negatively for each type of rule since they can have different interfaces)

File: iati/core/rulesets.py [hunk 4]
```diff
@@ -23,17 +23,13 @@ def locate_constructor_for_rule_type(rule_type):
     """Locate the constructor for specific rule types.
 
     Args:
-        rule_type (str): The name of the type of Rule to identify the class for.
+    rule_type (str): The name of the type of Rule to identify the class for.
 
     Returns:
-        Rule implementation: A constructor for a class that inherits from Rule.
+    Rule implementation: A constructor for a class that inherits from Rule.
```
Comment (hayfield) [inline iati/core/rulesets.py]: The return type is `type`.

File: iati/core/rulesets.py [hunk 5]
```diff
@@ -67,35 +63,38 @@ def __init__(self, ruleset_str):
...
+
+        """
         try:
             jsonschema.validate(self.ruleset, iati.core.default.ruleset_schema())
         except jsonschema.ValidationError:
             raise ValueError
 
-    def set_rules(self):
+    def _set_rules(self):
         """Set the Rules of the Ruleset."""
```
Comment (hayfield) [inline iati/core/rulesets.py]: There are no args or return type. As such, it is somehat unclear from the docstring alone what this function is doing. Suggest an expansion of the docstring (while avoiding implementation details!)

File: iati/core/rulesets.py [hunk 6]
```diff
@@ -142,12 +144,7 @@ def _normalize_xpath(self, path):
         return self.xpath_base + '/' + path
 
     def _normalize_xpaths(self):
-        """Normalize xpaths by combining them with `xpath_base`.
-
-        Todo:
-            Discuss appropriate edge cases and what is going on here.
-
-        """
+        """Normalize xpaths by combining them with `xpath_base`."""
```
Comment (hayfield) [inline iati/core/rulesets.py]: Based on earlier confusion, probably worth expanding the docstring to detail why this may want overriding in child classes (when there is no `paths` attribute and such)

File: iati/core/rulesets.py [hunk 7]
```diff
@@ -164,10 +161,7 @@ def _valid_rule_configuration(self, case):
             ValueError: When the case is not valid for the type of Rule.
 
         """
-        try:
-            ruleset_schema_section = self._ruleset_schema_section()
-        except AttributeError:
-            raise
+        ruleset_schema_section = self._ruleset_schema_section()
```
Comment (hayfield) [inline iati/core/rulesets.py]: Since it's only used once and doesn't add a lot to verbosity, it may be clearer to call `seld._ruleset_schema_section()` directly in the `jsonschema.validate()` call.

File: iati/core/rulesets.py [hunk 8]
```diff
@@ -232,13 +227,17 @@ def __init__(self, xpath_base, case):
         super(RuleAtLeastOne, self).__init__(xpath_base, case)
 
     def is_valid_for(self, dataset):
-        """Check `dataset.xml_tree` has at least one instance of a given case for an Element.
+        """Check dataset has at least one instance of a given case for an Element.
```
Comment (hayfield) [inline iati/core/rulesets.py]: dataset -> Dataset

File: iati/core/rulesets.py [hunk 9]
```diff
@@ -232,13 +227,17 @@ def __init__(self, xpath_base, case):
...
 
         Args:
-            dataset.xml_tree: an etree created from an XML dataset.
+            dataset (iati.core.Dataset): An IATI Dataset object.
```
Comment (hayfield) [inline iati/core/rulesets.py]: Out of the surrounding context, this isn't entirely clear. Would suggest something that is more standalone such as `The Dataset that is to have its validity checked against the Rule.` (but tweaked slightly to be less clunky)

File: iati/core/rulesets.py [hunk 10]
```diff
@@ -232,13 +227,17 @@ def __init__(self, xpath_base, case):
...
 
         Returns:
-            Boolean value that changes depending on whether the case is found in the dataset.xml_tree.
+            bool: Changes depending on whether the case is found in the dataset.
+
+        Raises:
+            AttributeError: When an argument is given that is not a dataset object.
```
Comment (hayfield) [inline iati/core/rulesets.py]: Based on the description, a `TypeError` should be raised rather than an `AttributeError`.

File: iati/core/rulesets.py (hunk 9 above)
Comment (hayfield) [inline iati/core/rulesets.py]: The same goes for other cases like this.

File: iati/core/rulesets.py (hunk 10 above)
Comment (hayfield) [inline iati/core/rulesets.py]: The same goes for other cases like this.

File: iati/core/rulesets.py [hunk 11]
```diff
@@ -271,18 +270,41 @@ def is_valid_for(self, dataset):
         """Assert that the date value of `less` is older than the date value of `more`.
 
         Args:
-            dataset.xml_tree: an etree created from an XML dataset.
+            dataset (iati.core.Dataset): An IATI Dataset object.
 
         Return:
-            A boolean value. If `less` is older than `more`, return `True`.
+            bool: Changes depending on whether `less` is older than `more`.
```
Comment (hayfield) [inline iati/core/rulesets.py]: `Changes depending on whether` isn't entirely clear on which way round the comparison is made.

Comment (hayfield) [inline iati/core/rulesets.py]: older than -> chronologically before

See https://github.com/IATI/IATI-Rulesets/issues/26

File: iati/core/rulesets.py [hunk 12]
```diff
@@ -271,18 +270,41 @@ def is_valid_for(self, dataset):
...
         more_date = dataset.xml_tree.xpath(self.more)
-        earlier_date = time.strptime(less_date[0], '%Y-%m-%d')
-        later_date = time.strptime(more_date[0], '%Y-%m-%d')
+        earlier_dates = list()
+        later_dates = list()
+
+        for date in less_date:
+            if date == 'NOW':
+                earlier_dates.append(datetime.strptime(datetime.today(), '%Y-%m-%d'))
+            earlier_dates.append(datetime.strptime(date[0:10], '%Y-%m-%d'))
```
Comment (hayfield) [inline iati/core/rulesets.py]: `[0:10]` -> `[:10]`

File: iati/core/rulesets.py [hunk 13]
```diff
@@ -271,18 +270,41 @@ def is_valid_for(self, dataset):
...
-            A boolean value. If `less` is older than `more`, return `True`.
+            bool: Changes depending on whether `less` is older than `more`.
+
+        Raises:
+            AttributeError: When an argument is given that is not a dataset object.
+            ValueError: When a date is given that is not in the correct xsd:date format.
+            XPathEvalError: When no valid xpath argument is given.
+
+        Note:
+            `date` restricted to 10 characters in order to exclude possible timezone values but suboptimal solution.
```
Comment (hayfield) [inline iati/core/rulesets.py]: Because of this implementation detail, a string shorter than 10 characters should be tested.

File: iati/core/rulesets.py (hunk 11 above)
Comment (hayfield) [inline iati/core/rulesets.py]: The same goes for other cases like this.

File: iati/core/rulesets.py [hunk 14]
```diff
@@ -232,13 +227,17 @@ def __init__(self, xpath_base, case):
...
-            dataset.xml_tree: an etree created from an XML dataset.
+            dataset (iati.core.Dataset): An IATI Dataset object.
 
         Returns:
-            Boolean value that changes depending on whether the case is found in the dataset.xml_tree.
+            bool: Changes depending on whether the case is found in the dataset.
+
+        Raises:
+            AttributeError: When an argument is given that is not a dataset object.
+            XPathEvalError: When no valid xpath argument is given.
```
Comment (hayfield) [inline iati/core/rulesets.py]: Assuming `XPathEvalError` is something provided by lxml... a module-level TODO should be added to consider whether this implementation detail can be contained.

File: iati/core/rulesets.py (hunk 13 above)
Comment (hayfield) [inline iati/core/rulesets.py]: (other than the special case of `NOW`)

File: iati/core/rulesets.py (hunk 14 above)
Comment (hayfield) [inline iati/core/rulesets.py]: `no valid xpath argument` - The function doesn't take an argument called `xpath`. As such, it is unclear what this is referring to.

Comment (hayfield) [inline iati/core/rulesets.py]: The same goes for other cases like this.

File: iati/core/rulesets.py [hunk 15]
```diff
@@ -430,19 +478,23 @@ def is_valid_for(self, dataset):
...
-            dataset.xml_tree: an etree created from an XML dataset.
+            dataset (iati.core.Dataset): An IATI Dataset object.
 
         Returns:
-            A boolean value. If the `path` string starts with the `start` string, return `True`.
+            bool: Changes depending on whether the `path` text starts with the value of `start`.
+
+        Raises:
+            AttributeError: When an argument is given that is not a dataset object.
+            IndexError: When no valid xpath argument is given.
```
Comment (hayfield) [inline iati/core/rulesets.py]: This description sounds like it should be a `ValueError`

File: iati/core/tests/test_rulesets.py [hunk 16]
```diff
@@ -4,6 +4,7 @@
 import iati.core.rulesets
 import iati.core.resources
 import iati.core.tests.utilities
+import lxml
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: `lxml` is a library dependency, so should be above the `iati` imports

File: iati/core/tests/test_rulesets.py [hunk 17]
```diff
@@ -142,29 +142,26 @@ def test_rule_class_cannot_be_instantiated_directly_with_name(self):
 class TestRuleSubclasses(object):
     """A container for tests relating to all Rule subclasses."""
 
-    rule_constructors = list(map(iati.core.rulesets.locate_constructor_for_rule_type, iati.core.rulesets._VALID_RULE_TYPES))
-    """A list of constructors for the various types of Rule."""
+    @pytest.fixture(params=[
+        iati.core.rulesets.locate_constructor_for_rule_type(rule_type) for rule_type in iati.core.rulesets._VALID_RULE_TYPES
+    ])
+    def rule_constructor(self, request):
+        """List constructors for the various types of Rule."""
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: A particular fixture is only a single thing at any given time. This docstring sounds slightly too this-is-an-array-like.

File: iati/core/tests/test_rulesets.py [hunk 18]
```diff
@@ -142,29 +142,26 @@ def test_rule_class_cannot_be_instantiated_directly_with_name(self):
...
-    @pytest.mark.parametrize("rule_constructor", rule_constructors)
     @pytest.mark.parametrize("xpath_base", iati.core.tests.utilities.find_parameter_by_type(['str'], False))
     def test_rule_init_invalid_xpath_base(self, rule_constructor, xpath_base):
-        """Check that a Rule cannot be created when xpath_base is not a string.
-
-        Todo:
-            Ensure the case is valid - an empty dictionary isn't.
-        """
+        """Check that a Rule cannot be created when xpath_base is not a string."""
         case = dict()
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: `case` is invalid. As such, it cannot be asserted that it is the invalid `xpath_base` that is causing the `ValueError` to be raised.

One should be valid, the other invalid.

File: iati/core/tests/test_rulesets.py [hunk 19]
```diff
@@ -187,14 +183,38 @@ class RuleSubclassTestBase(object):
...
+        """Return valid xpath_base."""
+        return '//root_element'
+
+    @pytest.fixture
+    def empty_xpath_base(self):
+        """Empty_xpath_base."""
+        return ''
+
+    @pytest.fixture
+    def basic_rule(self, rule_constructor, valid_case, xpath_base):
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: Other fixtures are named `rule_[xxx]` - this is inconsistent and should be brought in line with the others.

File: iati/core/tests/test_rulesets.py [hunk 20]
```diff
@@ -187,14 +183,38 @@ class RuleSubclassTestBase(object):
...
         """Instantiate a basic Rule subclass."""
-        xpath_base = 'an xpath'
         return rule_constructor(xpath_base, valid_case)
 
     @pytest.fixture
+    def rule(self, rule_constructor, xpath_base, case_for_is_valid_for):
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: The name of this should be classified more specifically - `rule_[something-meaningful]`.

File: iati/core/tests/test_rulesets.py [hunk 21]
```diff
@@ -187,14 +183,38 @@ class RuleSubclassTestBase(object):
...
+        """Instantiate a rule with specific cases for checking the `is_valid_for` function."""
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: `Instantiate` not required - because it's a fixture, it is the Rule.

Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: The same goes for other similar wordings.

File: iati/core/tests/test_rulesets.py [hunk 22]
```diff
@@ -187,14 +183,38 @@ class RuleSubclassTestBase(object):
...
+
+    @pytest.fixture
+    def rule_empty_xpath_base_and_paths(self, rule_constructor, empty_xpath_base, empty_path_case):
+        """Instantiate a basic Rule subclass with an empty xpath_base and empty path case."""
+        return rule_constructor(empty_xpath_base, empty_path_case)
+
+    @pytest.fixture
     def rule_constructor(self, rule_type):
-        """Return the constructor for the current Rule type."""
+        """Instantiate the current Rule type."""
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: There is a fixture in `TestRuleSubclasses` with the same name. May be worth adding a `Note:` to one or both to clarify what the difference is.

File: iati/core/tests/test_rulesets.py [hunk 23]
```diff
@@ -220,32 +240,44 @@ def test_rule_paths(self, basic_rule):
                 assert path.startswith(basic_rule.xpath_base)
 
     def test_rule_invalid_case(self, rule_constructor, invalid_case):
-        """Check that a rule cannot be instantiated when the case is invalid.
-
-        Todo:
-            Add tests with empty path AND empty xpath_base.
-        """
+        """Check that a rule cannot be instantiated when the case is invalid."""
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: rule -> Rule

Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: Here and in other similar cases when referring to Rules in this sort of manner.

File: iati/core/tests/test_rulesets.py [hunk 24]
```diff
@@ -220,32 +240,44 @@ def test_rule_paths(self, basic_rule):
...
+    @pytest.mark.parametrize("junk_data", iati.core.tests.utilities.find_parameter_by_type([], False))
+    def test_is_valid_for_raises_error_on_non_permitted_argument(self, rule, junk_data):
+        """Check that a given rule returns expected error when passed an argument that is not a dataset."""
+        with pytest.raises(AttributeError):
+            rule.is_valid_for(junk_data)
+
+    def test_is_valid_for_raises_error_when_passed_an_etree(self, rule):
+        """Check that an error is raised if an etree is given as an argument instead of a dataset."""
+        with pytest.raises(AttributeError):
+            rule.is_valid_for(iati.core.resources.load_as_tree(iati.core.resources.get_test_data_path('valid_atleastone')))  # use more generic dataset
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: Should the comment be a TODO? If so, it should be fixed. If not, it should either be moved and clarified; or removed.

File: iati/core/tests/test_rulesets.py [hunk 25]
```diff
@@ -220,32 +240,44 @@ def test_rule_paths(self, basic_rule):
...
+
+    def test_empty_xpath_base_rule_init_normalised_paths(self, rule_empty_xpath_base):
+        """Check that a rule with an empty xpath_base can be instantiated correctly with normalised `paths`."""
+        # Exclude rules without `paths`.
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: Why are they being excluded?

File: iati/core/tests/test_rulesets.py [hunk 26]
```diff
@@ -220,32 +240,44 @@ def test_rule_paths(self, basic_rule):
...
+        if 'paths' in dir(rule_empty_xpath_base):
+            for path in rule_empty_xpath_base.paths:
+                assert path.startswith(rule_empty_xpath_base.xpath_base)
+
+    def test_empty_xpath_base_empty_path_is_valid_for(self, rule_empty_xpath_base_and_paths, valid_dataset):
+        """Check that a rule with an empty xpath_base and empty paths raises error."""
+        # RuleStartswith raises an IndexError instead.
+        with pytest.raises((lxml.etree.XPathEvalError, IndexError)):
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: A test should be checking for only a single type of error.

File: iati/core/resources/test_data/202/valid_dateorder.xml [hunk 27]
```diff
@@ -4,5 +4,9 @@
   <planned-disbursement type="1">
```
Comment (hayfield) [inline iati/core/resources/test_data/202/valid_dateorder.xml]: The rule name has an underscore in its name. Consistency with the name of test files would aid readability slightly.

File: iati/core/resources/test_data/202/invalid_dateorder_format.xml [hunk 28]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0"?>
+
+<root_element>
+  <planned-disbursement type="1">
+   <period-start iso-date="01-12-2007"/><!-- Dates are not in the correct xsd:date format -->
+   <period-end iso-date="02-11-2005"/>
+  </planned-disbursement>
```
Comment (hayfield) [inline iati/core/resources/test_data/202/invalid_dateorder_format.xml]: There is no `NOW` in this file.

File: iati/core/resources/test_data/202/valid_dateorder.xml [hunk 29]
```diff
@@ -4,5 +4,9 @@
...
    <period-start iso-date="2014-01-01" /><!-- Dates are in the correct order -->
    <period-end iso-date="2014-12-31" />
+   <period-start iso-date="2017-01-01" /><!-- Dates are in the correct order -->
```
Comment (hayfield) [inline iati/core/resources/test_data/202/valid_dateorder.xml]: # MAJOR PROBLEM

This should cause the `{'less': 'planned-disbursement/period-start/@iso-date', 'more': 'planned-disbursement/period-end/@iso-date'}` test to fail - the start date of `2017-01-01` is after the end date of `2014-12-31`.

File: iati/core/tests/test_rulesets.py [hunk 30]
```diff
@@ -387,12 +378,16 @@ def valid_dataset(self):
         return iati.core.tests.utilities.DATASET_FOR_DATEORDER_RULE_VALID
 
     @pytest.fixture(params=[
-        {"less": "planned-disbursement/period-start/@iso-date", "more": "planned-disbursement/period-end/@iso-date"}
+        {'less': 'planned-disbursement/period-start/@iso-date', 'more': 'planned-disbursement/period-end/@iso-date'}
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: Needs a range of separate tests writing. Based on the params to `valid_case`, would expect something in the region of 12-20.

File: iati/core/tests/test_rulesets.py [hunk 31]
```diff
@@ -408,6 +403,64 @@ def test_rule_paths_more(self, basic_rule):
...
+        """Invalid dataset for this Rule."""
+        return iati.core.tests.utilities.DATASET_FOR_DEPENDENT_RULE_INVALID
+
+    @pytest.fixture
+    def valid_dataset(self):
+        """Return valid dataset for this Rule."""
+        return iati.core.tests.utilities.DATASET_FOR_DEPENDENT_RULE_VALID
+
+    @pytest.fixture(params=[
+        {"paths": ["transaction/provider-org", "location/point"]}
```
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: # MAJOR PROBLEM

There are a number of valid cases passed in to `valid_case` that are untested.

`{'paths': ['']},  # empty path`, `{'paths': ['path_1']},  # single path`, `{'paths': ['path_1', 'path_1']}  # duplicate paths`

For each of these, there should be a dataset that is valid against the Rule, plus a dataset that is invalid against the Rule.

Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: The same goes for other specific types of Rule where there are multiple cases passed into `valid_case`.

Heck, `valid_case` is a fixture, so one way to deal with this is to use the values directly, and for each case create a valid dataset and an invalid dataset.

File: iati/core/default.py [hunk 32]
```diff
@@ -120,14 +120,17 @@ def ruleset(version=None):
 
     return iati.core.Ruleset(ruleset_str)
 
+
 def ruleset_schema(version=None):
+    """Todo: Add docstring."""
```
Comment (allthatilk) [inline iati/core/default.py]: done

File: iati/core/resources.py [hunk 33]
```diff
@@ -317,41 +293,43 @@ def get_path_for_version(path, version=None):
...
 
-def load_as_bytes(path):
-    """Load a resource at the specified path into a bytes object.
+
+def load_as_dataset(path):
+    """Load a resource at the specified path into a dataset.
 
     Args:
-        path (str): The path to the file that is to be read in.
+    path (str): The path to the file that is to be read in.
```
Comment (allthatilk) [inline iati/core/resources.py]: done

File: iati/core/resources/test_data/202/valid_dateorder.xml (hunk 27 above)
Comment (allthatilk) [inline iati/core/resources/test_data/202/valid_dateorder.xml]: I haven't done this as the original rule names are inconsistently underscored.

File: iati/core/resources/test_data/202/valid_dateorder.xml (hunk 1 above)
Comment (allthatilk) [inline iati/core/resources/test_data/202/valid_dateorder.xml]: Why? They all get sliced off in the same way. It would be a redundant addition.

File: iati/core/rulesets.py (hunk 4 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/rulesets.py (hunk 5 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: It doesn't return anything but otherwise done.

Comment (allthatilk) [inline iati/core/rulesets.py]: It doesn't return anything and has no args so... not sure what else to add tbh.

File: iati/core/rulesets.py (hunk 6 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/rulesets.py (hunk 7 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/rulesets.py (hunk 8 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/rulesets.py (hunk 9 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/rulesets.py (hunk 10 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: Okay, so should I change the description or wrap all functions in a try-catch? Currently going for the former.

File: iati/core/rulesets.py (hunk 14 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: done

Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/rulesets.py (hunk 11 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: As this was changed from the original from containing too much implementation detail, I have simply changed it  to include implementation detail as I'm not sure how else to reach the desired specificity here.

Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/rulesets.py (hunk 12 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/rulesets.py (hunk 15 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: Updated error description.

File: iati/core/tests/test_rulesets.py (hunk 16 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: done

File: iati/core/tests/test_rulesets.py (hunk 17 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: Fixed

File: iati/core/tests/test_rulesets.py (hunk 19 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: done

File: iati/core/tests/test_rulesets.py (hunk 20 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: done

File: iati/core/tests/test_rulesets.py (hunk 21 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: done

File: iati/core/tests/test_rulesets.py (hunk 22 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: done

File: iati/core/tests/test_rulesets.py (hunk 23 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: done

File: iati/core/tests/test_rulesets.py (hunk 24 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: done

File: iati/core/tests/test_rulesets.py (hunk 25 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: Because the test is specific to Rules that include `paths`. If a Rule has no `paths` by design (RuleDateOrder), this test will incorrectly error.

File: iati/core/tests/test_rulesets.py (hunk 26 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: It can't by design due to the differences in Rules.

File: iati/core/resources/test_data/202/invalid_unique.xml [hunk 34]
```diff
@@ -1,6 +1,8 @@
 <?xml version="1.0"?>
 
 <root_element>
-  <iati-identifier>This is not a unique identifier</iati-identifier><!-- iati-identifier is replicated -->
+  <iati-identifier>This is not a unique identifier</iati-identifier><!-- elements are replicated -->
```
Comment (allthatilk) [inline iati/core/resources/test_data/202/invalid_unique.xml]: Then it's valid.

Comment (allthatilk) [inline iati/core/resources/test_data/202/invalid_unique.xml]: Then it's valid.

File: iati/core/resources/test_data/202/valid_unique.xml (hunk 2 above)
Comment (allthatilk) [inline iati/core/resources/test_data/202/valid_unique.xml]: Agreed. Will change with the general overhaul of test data.

File: iati/core/rulesets.py (hunk 5 above)
Comment (allthatilk) [inline iati/core/rulesets.py]: done

File: iati/core/resources/test_data/202/invalid_unique.xml (hunk 34 above)
Comment (allthatilk) [inline iati/core/resources/test_data/202/invalid_unique.xml]: Hmmm, I should change the comment to specify that.

File: iati/core/resources.py [hunk 35]
```diff
@@ -317,41 +293,43 @@ def get_path_for_version(path, version=None):
...
 
     """
-    dataset_str = load_as_string(path)
-    return iati.core.Dataset(dataset_str)
+    return pkg_resources.resource_string(PACKAGE, path)
 
-def load_as_bytes(path):
-    """Load a resource at the specified path into a bytes object.
+
+def load_as_dataset(path):
```
Comment (allthatilk) [inline iati/core/resources.py]: done

File: iati/core/tests/test_rulesets.py (hunk 18 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: done. Moved to RuleSubclassTestBase.

File: iati/core/tests/test_rulesets.py (hunk 31 above)
Comment (allthatilk) [inline iati/core/tests/test_rulesets.py]: valid_case is only for testing whether the Rule instantiates. It's not possible for all of them to validate a dataset, for example {'paths': ['']} is a valid path for instantiation but won't always return True for valid data because it won't always create a valid XPath search (this problem of empty paths applies to all Rules and is tested elsewhere).

File: iati/core/rulesets.py (hunk 10 above)
Comment (hayfield) [inline iati/core/rulesets.py]: Assuming the description is the desired situation in which an error occurs, I'd suggest wrapping the relevant statement in try-except, catching the `AttributeError` and raising a `TypeError`.

File: iati/core/tests/test_rulesets.py (hunk 25 above)
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: An explanation like that would be helpful in the docstring

File: iati/core/tests/test_rulesets.py (hunk 26 above)
Comment (hayfield) [inline iati/core/tests/test_rulesets.py]: The API for the different Rule subclasses should be consistent.

This relates somewhat to the TODO about maybe hiding `lxml.etree.XPathEvalError`, but would suggest making that the desired exception (so either `try-except`ing ths `IndexError` and raising the other one; or (probably better) making a lxml call that raises the exception itself)
